feat: add ability to use function as config (#913)
okonet/lint-staged: README.md

going to execute `eslint` and if it exits with `0` code, it will execute `prettier --write` on all staged `*.js` files.

## Using JS configuration file

> Writing `lint-staged.config.js` file (or, other JS file which needs to be mentioned in CLI command through `c` flag) is the most powerful way to configure `lint-staged`. From the configuration file, you can export a function which gets an array of all staged file paths as parameter and returns command(s) as string or array of string. This function can be either sync or async. Signature of this function:

Reviewer: ```suggestion
Writing the configuration file in JavaScript is the most powerful way to configure _lint-staged_ (`lint-staged.config.js`, similar, or passed via `--config`). From the configuration file, you can export either a single function, or an object.

If the `exports` value is a function, it will receive an array of all staged filenames. You can then build your own matchers for the files, and return a command string, or an array or command strings. These strings are considered complete and should include the filename arguments, if wanted.

It the `exports` value is an object, its keys should be glob matches (like in the normal non-js config format). The values can either be like in the normal config, or then individual functions like described above. Instead of receiving all matched files, the functions in the exported object will only receive the staged files matching the corresponding glob key.
```

Reviewer: the `similar` word could actually link to the part about using cosmiconfig:

```md
[suggestion](https://github.com/okonet/lint-staged/README.md#configuration)
```

Author: This is much better, but there's a typo in the last line. `It` should be `If`.

Reviewer: Good catch, fixed!

Reviewer: I also put in the similar-link into the suggestion.

```ts
(filenames: string[]) => string | string[] | Promise<string | string[]>
```

> ### Example:

Reviewer: ```suggestion
### Example: Export a function to build your own matchers
```

```js
// lint-staged.config.js
const micromatch = require('micromatch')

module.exports = (allStagedFiles) => {
    const shFiles =  micromatch(allStagedFiles, ['**/src/**/*.sh']);
    if (shFiles.length) {
      return `printf '%s\n' "Script files aren't allowed in src directory" >&2`
    }
    const codeFiles = micromatch(allStagedFiles, ['**/*.js', '**/*.ts']);
    const docFiles = micromatch(allStagedFiles, ['**/*.md']);
    return [`eslint ${codeFiles.join(' ')}`, `mdl ${docFiles.join(' ')}`];
  }
```

> You can also supply configuration as an object to enjoy in-built filter mechanism like other configuration methods. One extra advantage in the case of JS config is that it is possible to define the task as a function, which has signature of the above function (The only difference is that task functions don't receive array of all staged file paths as parameter unless they have `*` filter pattern).

Reviewer 2: I'd provide an example for that too. Now it's not clear it's possible anymore.

Author: Existing examples of object based configuration are still there.

> You can also supply configuration as an object to enjoy in-built filter mechanism like other configuration methods. One extra advantage in the case of JS config is that it is possible to define the task as a function, which has signature of the above function (The only difference is that task functions don't receive array of all staged file paths as parameter unless they have `*` filter pattern).

Reviewer: ```suggestion
```

I would remove this, since the other suggestion includes the text.

going to execute `eslint` and if it exits with `0` code, it will execute `prettier --write` on all staged `*.js` files.

## Using JS configuration file

Writing the configuration file in JavaScript is the most powerful way to configure _lint-staged_ (`lint-staged.config.js`, [similar](https://github.com/okonet/lint-staged/README.md#configuration), or passed via `--config`). From the configuration file, you can export either a single function, or an object.

If the `exports` value is a function, it will receive an array of all staged filenames. You can then build your own matchers for the files, and return a command string, or an array or command strings. These strings are considered complete and should include the filename arguments, if wanted.

If the `exports` value is an object, its keys should be glob matches (like in the normal non-js config format). The values can either be like in the normal config, or individual functions like described above. Instead of receiving all matched files, the functions in the exported object will only receive the staged files matching the corresponding glob key.

> ```ts
> (filenames: string[]) => string | string[] | Promise<string | string[]>

Reviewer: I think I accidentally forgot the title for this, sorry about that:

```suggestion
### Function signature

The function can also be async:

```ts
(filenames: string[]) => string | string[] | Promise<string | string[]>
```

```

### Example: Use your own globs

> It's better to use function-based configuration if your use case is this.

Reviewer 2: I'd update the example instead of this paragraph

Author: What do you mean? Example of function based configuration is already in the documentation. Maybe, a link to that example can be added.

> It's better to use function-based configuration if your use case is this.

Reviewer: ```suggestion
It's better to use the [function-based configuration (seen above)](https://github.com/okonet/lint-staged/blob/b5245e3f1332a09cc8686dfaadc54d1d7b0ce79e/README.md#example-export-a-function-to-build-your-own-matchers), if your use case is this.

Reviewer: Oops, the link should be for master:

```suggestion
It's better to use the [function-based configuration (seen above)](https://github.com/okonet/lint-staged/README.md#example-export-a-function-to-build-your-own-matchers), if your use case is this.
```
